Use applyOrElse for PartialFunctions

szeiger opened the pull request:
This avoids evaluating the guards twice. It can be significantly
(25% in my benchmarks) slower in the worst case (a PF with a single
clause with very efficient guard that usually does not match) but can
have much larger performance gains in other scenarios.

All flatMap-based implementations suitable for lazy collection types
are replaced by a new `View.Collect` primitive that uses an Iterator
instead, which is considerably faster.

Fixes https://github.com/scala/bug/issues/10901

szeiger committed:
Use applyOrElse for PartialFunctions

This avoids evaluating the guards twice. It can be significantly
(25% in my benchmarks) slower in the worst case (a PF with a single
clause with very efficient guard that usually does not match) but can
have much larger performance gains in other scenarios.

All flatMap-based implementations suitable for lazy collection types
are replaced by a new `View.Collect` primitive that uses an Iterator
instead, which is considerably faster.

allanrenucci commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -416,7 +417,8 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
     // Manually buffer to avoid extra layer of wrapping with buffered
-    private[this] var hd: A = _
+    private[this] var hd: B = _
+    private[this] final val marker = ScalaRunTime.pfMarker // this should be static if there was a way to make it
```
Does this need to be a field? Why not a local variable in `hasNext`?

szeiger replied:
A local variable would have to be initialized on every call to `hasNext`. This requires reading an object's field which is slower than a regular instance field. Ideally it should be a static field.

retronym replied:
I would suggest defining `pfMarker` in `scala.runtime.Statics` and removing this field.

viktorklang commented on src/library/scala/collection/View.scala:
```diff
@@ -211,6 +211,12 @@ object View extends IterableFactory[View] {
 
+  /** A view that collects elements of the underlying collection. */
+  @SerialVersionUID(3L)
+  class Collect[+A, B](underlying: SomeIterableOps[A], pf: PartialFunction[A, B]) extends AbstractView[B] {
```
not final?

szeiger replied:
None of those classes are. More specific View types like IndexedSeqView have to be able to extend them.

allanrenucci reviewed (commented):
Should we update `PartialFunction` to use a local marker instead of a self returning function?
https://github.com/scala/scala/blob/80ee48b1880c8f817581a90f483b394273c237bc/src/library/scala/PartialFunction.scala#L195-L218

viktorklang commented on src/library/scala/util/Try.scala:
```diff
@@ -225,10 +226,20 @@ final case class Failure[+T](exception: Throwable) extends Try[T] {
-    try { if (pf isDefinedAt exception) Success(pf(exception)) else this } catch { case NonFatal(e) => Failure(e) }
-  override def recoverWith[U >: T](pf: PartialFunction[Throwable, Try[U]]): Try[U] =
-    try { if (pf isDefinedAt exception) pf(exception) else this } catch { case NonFatal(e) => Failure(e) }
+  override def recover[U >: T](pf: PartialFunction[Throwable, U]): Try[U] = {
```
Great catch, Stefan!

julienrf commented on src/library/scala/collection/immutable/LongMap.scala:
```diff
@@ -476,7 +476,7 @@ sealed abstract class LongMap[+T] extends AbstractMap[Long, T]
 
   def collect[V2](pf: PartialFunction[(Long, T), (Long, V2)]): LongMap[V2] =
-    flatMap(kv => if (pf.isDefinedAt(kv)) new View.Single(pf(kv)) else View.Empty)
+    LongMap.from(new View.Collect(coll, pf))
```
Why not builder-based implementations instead?

szeiger replied:
Yes, builder-based would probably be better.

szeiger replied:
I'll use https://github.com/scala/scala/pull/6776 (which I just merged) for this purpose.

szeiger committed:
Build speclib/instrumented.jar for partest

- Build automatically from sources instead of fetching a pre-built
  binary which can get out of date and needs to be updated manually

- Remove outdated build script (from the ant days)

szeiger commented:
I've added a commit to build speclib locally instead of using a pre-built binary. The old version was no longer compatible.

SethTisue commented:
/synch

szeiger commented:
See https://github.com/scala/bug/issues/7822 and https://github.com/scala/scala/commit/94166682ee for context for the speclib stuff.

szeiger commented:
Rebased and updated to use the reusable strict implementations. I've also moved `pfMarker` into `Statics`.

> Should we update `PartialFunction` to use a local marker instead of a self returning function?

The common wisdom so far was that a self-returning function is faster. My benchmarks show the opposite but they only test the collection-specific use-cases. I didn't want to make such a change without more data.
